### DISCUSSION TITLE
Which combinations of "output.target" and "output.library.type" are being frequently used in Webpack configuration?

### DISCUSSION THREAD

**Discussion by TokugawaTakeshi** [webpack/webpack#16020] [Q&A] [answered]
I suppose not all combinations of `output.target` and `output.library.type` are valid and/or meaningful. Which ones are meaningful and being frequently used?

### Case 1: Library for Browser JS planned to be imported by Webpack 

* Obviously, the **target** must be `web`
* **output.library.type** must be the ... `module`, I suppose, because using Webpack we don't need to take the functionality from the global scope (`window`).
* For TypeScript case: the `modules` should be set to `ES6` or above to enable the [tree shaking](https://webpack.js.org/guides/tree-shaking/) functionality on production mode

## Case 2: Web library which intended to be accessible via `script` tag

* Obviously, the **target** must be `web` 
* **output.library.type** must be the ... `window`, I suppose.

## Case 3: NodeJS library

Altgough the develpment team of Node.js seriously improved the ECMAScript modules support,
I suppose it is too early (for 2022) to tell about ECMAScript modules as main modules of Node.js.

* Obviously, the `target` must be the `node`. 
* **output.library.type** must be the ... `commonjs`, I suppose, but there are similar `commonjs2` and `commonjs-module` there. It will take a lot of time to explore the differences in details, so I hope on your reccomendations.

### Case 4: The library which could be used by both BrowserJS and Node.js

* The The [Authoring Libraries](https://webpack.js.org/guides/author-libraries/) guide does not chage the default `web` value of the `target`, thus, it will be `web` even if the library will be used in Node.js environment.
* The [Authoring Libraries](https://webpack.js.org/guides/author-libraries/) guide reccomends "umd" as `library.type`. But... will tree shaking available with this type at least, for web. Afraid, no. Is so, it's better to use multicompilation with different target environments.

But: it was the rare case where UMD was usefull for me.
I tried to build the JavaScript for... Pug pre-processor, which has internal JavaScript environment. It does not understand neither `import` nor `require`, so `umd` was a solution there.

## Remarks

* `libraryTarget` is planned to drop in the future.
*  I am asking this question here after [has not been answered on Stack Overflow](https://stackoverflow.com/questions/72723248/which-combinations-of-output-target-and-output-library-type-are-being-freque) even with bounty.

**Comment by vankop** [comment, marked as answer]
`target` is more about [chunkformat](https://webpack.js.org/configuration/output/#outputchunkformat) and [chunkloading](https://webpack.js.org/configuration/output/#outputchunkloading) options in this case

`library.type` about exports format..
_________

technically all combinations are valid... In most cases now esm libraries are used
